### PR TITLE
Xenobio Addition

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports half of my own PR from TG: https://github.com/tgstation/-tg-station/pull/16694

- Adds in support for self-recharging cells; they regain 25% of their recharge rate every `process` (this is far far slower than putting them in a charger)
- Slime cells self-recharge by default

:cl: Fox McCloud
add: Slime batteries now self-recharge.
/:cl:

### CODE REVIEW

**Comment by Bigtimetripod** [comment]
Oh damn, here comes the wave of "NAW YUU CAN TURN SEC INTO WALKIN LING ARMY, NERF PLOX"

Or did that get removed... xenobio has gone thru so much changes lately in starting to loose track.

Oo.. self recharging cells... interesting. Xenobios effectivness just went up by.. 48%. Imagine, power outage, idiots eff up tesla (can happen) and APC are running dry... Xenobio to rescue! Im thinking the recharge rate is high enough to keep up RnD computer meaning it can print out more cells for departments that really need it and get singulo up and running.

Back to ling army topic... since its hard to get rainbows (late-round stuff) ye will not see much of sec/command transfering themself into lings. Maybe set somekind of SOP? Dunno, topic for higher minds.

**Comment by FlattestGuitar** [comment]
There already is a science SOP, which every single science player ardently ignores. Currently everything about xenobio makes it turn into a horribly chaotic mess, but it stays in the room 90% of the time, so not that much of a problem. 

This isn't really the topic of this PR though. 

I'm pretty sure if the cells have a recharge rate this high they may be a bit too powerful. Does that mean you can make infinite stunprods? What about using them in APCs, will that disable the need for any external power most of the time?

**Comment by Fox-McCloud** [comment]
The current recharge rate is very low; you're talking a mere +375 charge every 2 seconds (for slime cells)---it's not going to keep the power on, indefinitely, but it can be of use to keep things recharged, on the fly, that only see 1-2 uses and that's it---like defibs, for example.

**Comment by Regen1** [comment]
:-1: :100: No please god no

Can we please NOT let people create their favorite mob just to take control over it themselves? This is literally as bad as just suiciding to join as an NPC, it's just a game mechanic that justifies it.

>Play the role you've chosen

Who in their right mind would transfer their bloody soul into an unknown creature?

**Comment by tigercat2000** [comment]
I'm going to with @Regen1 here.
Allowing players to deliberately make themselves into a simple animal of their choice is just god awful, and asking for trouble and metagaming.

**Comment by Spacemanspark** [comment]
I think xenobiology has had enough changes and buffs.
:-1:

**Comment by Crazylemon64** [comment]
+375 charge is sufficient to let a borg do whatever it wants and not fear ever running out, provided that it isn't spamming a borg RCD - as a cell in a mech, it'll let you take 37 steps every 2 seconds, of which you can only probably do 3-5.

**Comment by FalseIncarnate** [comment]
I like the self-recharging cells support, pretty much makes a power equivalent to the experimental welding tool. Only real issue with it is that the uses are fairly limited since you can't swap batteries in the most common user of them: guns. Still, a self-charging APC or defib unit will be a handy tool for science to hook medbay up with. A self charging SMES might be interesting too, if those work with these.

I can't say I'm a huge fan of the new potion though. Perhaps allow you to transfer the consciousness between two different simple animals, like to transfer Ian's mind into a diyab or something, or between two human mobs (swap minds with your co-worker), but to allow the scientist to effectively throw himself into the newly spawned xeno queen or (as people have been worrying about) changeling would be a bit of borderline self-antag behavior.

I understand that it would largely be a one-way transference, since most mobs lack the capability to manipulate the items necessary to swap back, but that'd just encourage people to swap with the strongest and most versatile ones they can get (xenos, changelings, etc).

**Comment by Dumbdumn5** [comment]
Eh, I actually enjoy the new potion, mainly because we already have an extract that turns you into a slime. If there could be a blacklist on the more hostile mobs (xenos, changelings, etc.) I'd hope that'd at least change something. Everything else seems fine enough.

**Comment by TheDZD** [comment]
I like the idea of self-recharging batteries, but as long as players can make themselves into changelings, or other such mobs that are too absurdly powerful to have be allied with the crew, I'm going to have to vehemently disagree with the transference potion.

**Comment by Regen1** [comment]
@TheDZD this isn't just a "Balance" problem anymore, it's about players leaving their character to play a fucking simplemob

Let me put it differently, we would ban players from science if they went to xenobio, made sentient potions and a bunch of mobs, suicided and asked the other guy in xenobio to use the sentient potion on whatever animal.

This is just a game mechanic that somehow justifies something we would ban for.

**Comment by Iamgoofball** [comment]
@Regen1 @tigercat2000 Realism is not going to overpower gameplay here. There's many many other actual balance issues you can argue with, like the headslug thing, or becoming hostile mobs in specific and using it to murderbone. Use those as argument points. No one in power is going to listen to you guys on the github if the only argument is "muh reels".

Your character is just a shell you inhabit. Why does it matter if Randomname McRandomson turns himself into a corgi? Why does it matter? Does them being a simple mob ruin YOUR fun in the game? Does a person playing a job that is already isolated from the station completely, turning themselves into a harmless animal, and fucking about not doing anything really cause a serious balance problem?

@Spacemanspark "lets just never change it ever again"
thats totally not the thinking that results in stagnant gameplay mechanics and departments that havent been updated in 5 years

**Comment by tigercat2000** [comment]
@Iamgoofball I wasn't agreeing with his second point. Fuck balancing code around "realism", that's just stupid. I was agreeing with him because of letting people become any simple animal they so please is a very bad thing, *especially* with the tensions already around the fact that xenobiology can create little slave changelings to fight the antagonist.

**Comment by ProperPants** [comment]
@tigercat2000 About the headslug thing, can't somebody just blacklist them? It's probably not as simple as _just_, but if it's such a catastrophic issue, shouldn't somebody be on it by now? Maybe someone is and I just don't know, but it seems theoretically simple (but of course that doesn't mean it's practically simple).

**Comment by Fox-McCloud** [comment]
>and metagaming.

Because someone in the round becoming a simple mob has more potential for metagaming than a ghost observer who has seen what's been going on for the past 15+ minutes.

**Comment by Crazylemon64** [comment]
@Fox-McCloud In this case, the person becoming the simple mob has 100% control over what choices they want to become - a ghost observer lacks this agency. Leading to, possibly, someone who really wants to be a swarmer, to create simplemobs till they get a swarmer, then becoming one and starting an antag event on their own initiative, all without interacting with anyone prior - we call this "Self antagging", no?

**Comment by Fox-McCloud** [comment]
And then they get banned for griefing if they weren't an antag prior to turning themselves into one--the potion doesn't magically confer antag status on the person just because they have it; if they're a non-antag and they become something like that, then they're griefing.

That's no different than the person who goes scientist, and wants create an antag event of their own initiative, so they make a toxin's bomb and blow up the bridge/bar/what have you. We don't talk about removing Toxins because of that; we ban the person for griefing.

**Comment by Draconic-saint** [comment]
I'm not a fan of the current transference potion. 

My biggest issue is the changeling headcrab.

Meta-/Powergamers WILL turn themselves into a changeling, and they can validhunt with their powers, cause unneeded chaos by kinda being an antag, but not really, and just in general cause more shit that we need to keep track of.

And if you're a traitor or any other kind of antag that benifits from having extra firepower, then it maybe makes sense, hell it could add some extra spice. Maybe not for all antags though. I really don't want any changelings in my cult.

At least blacklist the changeling headcrab for non antags, if that distinction is possible. Maybe some other mobs as wel, but I do feel that non-antags shouldn't be able to acces changeling powers like that.

I know that some people aren't a fan of players joining and "being useless". I guess this also falls under that category, but that's an aspect of the potion I don't have an issue with.

Griefers gonna grief, and they have much efficienter tools than this to do it. 

Make an exception for the headcrab, and i'm game.
(You could always remove the headcrab as a gold slime spawnable mob, but this isn't the place for that discussion)

**Comment by Iamgoofball** [comment]
@tigercat2000 Oh, sorry. Yeah, I can see the problem with "brb injecting self into space carp, its valid to kill then :^)". My main problem was muh reels as the excuse.

**Comment by fludd12** [comment]
Maybe the transference potion in another PR, due to the controversy? I feel like nobody'd object to a buff to the rarely used yellow battery reaction.

**Comment by Fox-McCloud** [comment]
For reference, players can no longer receive headcrabs from gold slime cores: https://github.com/ParadiseSS13/Paradise/pull/4160

**Comment by tigercat2000** [comment]
Still really don't like allowing people to put themselves in simple animal xenomorphs/drones/goliaths.

**Comment by Fox-McCloud** [comment]
>Still really don't like allowing people to put themselves in simple animal xenomorphs/drones/goliaths. 

which is inherently no different than pulling in a ghost for the same, exact thing.

**Comment by tigercat2000** [comment]
@Fox-McCloud The difference is that you can have a reason beyond "a thing popped up and offered me a mob". Killing people, sneaking into places, or being a cute dog, or any other number of reasons, you get to decide to do so.

**Comment by Fox-McCloud** [comment]
And so do they; they get a pop-up informing them if they want to be X mob, which they can decline or accept----and as they're a ghost, they can generally suss out who is or is not the antag, even without antag hud (not to mention it only takes one ghost to have antag hud on for alllll the ghosts to know). Do they get to choose, exactly? No, they don't, but they still have a decent amount of agency and judgment over things.

**Comment by fludd12** [comment]
I can see this becoming an interesting weapon of last resort for miners, where instead of dying, they become the goliath that would have killed them.

**Comment by Keilden** [comment]
> I can see this becoming an interesting weapon of last resort for miners, where instead of dying, they become the goliath that would have killed them.

Captain Ginyu style?

**Comment by Rumiluntti** [comment]
Stop shoveling buffs into science for fuck's sake

**Comment by Regen1** [comment]
I am still heavily against the transference potion

**Comment by Iamgoofball** [comment]
Regen when are you going to provide a valid reason

**Comment by Crazylemon64** [comment]
Being able to choose your shell allows you to vet your form for maximum power, and as stated above, changeling headcrabs will let any xenobiologist get all the powers of a changeling without any personal risk or loss- since they can just morph back to their original form once they become a humanoid changeling, still have all the abilities of a normal human. This seems like it would inflame the already heavily present powergaming in science, wouldn't it?

**Comment by Regen1** [comment]
Who the fuck even is this guy?

Didn't github get an update? Can we please ban him already

**Comment by Fox-McCloud** [comment]
@Crazylemon64 

That hasn't been a thing for well over a week now

@Regen1 

Knock it off with the personal attacks.

**Comment by Regen1** [comment]
jesus christ get a grip

File: code/modules/research/xenobiology/xenobiology.dm [hunk 1]
```diff
@@ -223,6 +223,49 @@
...
+	icon_state = "bottle6"
+	origin_tech = "biotech=6"
+	var/prompted = 0
+	var/animal_type = SENTIENCE_ORGANIC
+
+/obj/item/slimepotion/transference/afterattack(mob/living/M, mob/user)
+	if(prompted || !ismob(M))
+		return
+	if(!isanimal(M) || M.ckey) //much like sentience, these will not work on something that is already player controlled
+		user << "<span class='warning'>[M] already has a higher consciousness!</span>"
```
Comment (tigercat2000) [inline code/modules/research/xenobiology/xenobiology.dm]: to_chat

File: code/modules/research/xenobiology/xenobiology.dm [hunk 2]
```diff
@@ -223,6 +223,49 @@
...
+		return ..()
+	if(M.stat)
+		user << "<span class='warning'>[M] is dead!</span>"
```
Comment (tigercat2000) [inline code/modules/research/xenobiology/xenobiology.dm]: to_chat

File: code/modules/research/xenobiology/xenobiology.dm [hunk 3]
```diff
@@ -223,6 +223,49 @@
...
+	user << "<span class='notice'>You drink the potion then place your hands on [SM]...</span>"
+
+
+	user.mind.transfer_to(SM)
+	SM.universal_speak = 1
+	SM.faction = user.faction
+	SM.sentience_act() //Same deal here as with sentience
+	user.death()
+	SM << "<span class='notice'>In a quick flash, you feel your consciousness flow into [SM]!</span>"
+	SM << "<span class='warning'>You are now [SM]. Your allegiances, alliances, and roles are still the same as they were prior to consciousness transfer!</span>"
```
Comment (tigercat2000) [inline code/modules/research/xenobiology/xenobiology.dm]: to_chat

File: code/modules/research/xenobiology/xenobiology.dm [hunk 4]
```diff
@@ -223,6 +223,49 @@
...
+
+	user << "<span class='notice'>You drink the potion then place your hands on [SM]...</span>"
+
+
+	user.mind.transfer_to(SM)
+	SM.universal_speak = 1
+	SM.faction = user.faction
+	SM.sentience_act() //Same deal here as with sentience
+	user.death()
+	SM << "<span class='notice'>In a quick flash, you feel your consciousness flow into [SM]!</span>"
```
Comment (tigercat2000) [inline code/modules/research/xenobiology/xenobiology.dm]: to_chat

File: code/modules/research/xenobiology/xenobiology.dm [hunk 5]
```diff
@@ -223,6 +223,49 @@
...
+	if(jobban_isbanned(user, ROLE_SENTIENT))
+		user << "<span class='warning'>Your mind goes blank as you attempt to use the potion.</span>"
+		return
+
+	prompted = 1
+	if(alert("This will permanently transfer your consciousness to [SM]. Are you sure you want to do this?",,"Yes","No")=="No")
+		prompted = 0
+		return
+
+	user << "<span class='notice'>You drink the potion then place your hands on [SM]...</span>"
```
Comment (tigercat2000) [inline code/modules/research/xenobiology/xenobiology.dm]: to_chat

File: code/modules/research/xenobiology/xenobiology.dm [hunk 6]
```diff
@@ -223,6 +223,49 @@
...
+		return ..()
+	if(M.stat)
+		user << "<span class='warning'>[M] is dead!</span>"
+		return ..()
+	var/mob/living/simple_animal/SM = M
+	if(SM.sentience_type != animal_type)
+		user << "<span class='warning'>You cannot transfer your consciousness to [SM].</span>" //no controlling machines
+		return ..()
+	if(jobban_isbanned(user, ROLE_SENTIENT))
+		user << "<span class='warning'>Your mind goes blank as you attempt to use the potion.</span>"
```
Comment (tigercat2000) [inline code/modules/research/xenobiology/xenobiology.dm]: to_chat

File: code/modules/research/xenobiology/xenobiology.dm [hunk 7]
```diff
@@ -223,6 +223,49 @@
...
+		return
+	if(!isanimal(M) || M.ckey) //much like sentience, these will not work on something that is already player controlled
+		user << "<span class='warning'>[M] already has a higher consciousness!</span>"
+		return ..()
+	if(M.stat)
+		user << "<span class='warning'>[M] is dead!</span>"
+		return ..()
+	var/mob/living/simple_animal/SM = M
+	if(SM.sentience_type != animal_type)
+		user << "<span class='warning'>You cannot transfer your consciousness to [SM].</span>" //no controlling machines
```
Comment (tigercat2000) [inline code/modules/research/xenobiology/xenobiology.dm]: to_chat

**Pull request merged**